From 54fd3924b8333f2583f0e05115c21fdef5b2d7ae Mon Sep 17 00:00:00 2001
From: Manjeet Rulhania <mrulhania@google.com>
Date: Wed, 30 Oct 2024 20:18:46 +0000
Subject: [PATCH] Role re-evaluation should respect ask every time choice by
 user

Role evalation respect USER_SET flag when the role is not
supposed to override user choices. Role granting logic
is missing ONE_TIME check, ONE_TIME flag should be considered
a user choice.

Bug: 355411348
Test: RolePermissionOverrideTest
Relnote: security bug fix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:71e09e5ad12d02b3129c7a5ce16f6de734fdd0da)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:6687e8ebe51d3d180dd926e418ec0681a4a79b9b)
Merged-In: I4146abff462ef4ed0e67efd04b4182bcf2c04d98
Change-Id: I4146abff462ef4ed0e67efd04b4182bcf2c04d98
---
 .../permissioncontroller/role/model/Permissions.java        | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/PermissionController/src/com/android/permissioncontroller/role/model/Permissions.java b/PermissionController/src/com/android/permissioncontroller/role/model/Permissions.java
index 4bacdcf28..c65bccc6e 100644
--- a/PermissionController/src/com/android/permissioncontroller/role/model/Permissions.java
+++ b/PermissionController/src/com/android/permissioncontroller/role/model/Permissions.java
@@ -253,7 +253,8 @@ public class Permissions {
         if (!wasPermissionOrAppOpGranted) {
             // If we've granted a permission which wasn't granted, it's no longer user set or fixed.
             newMask |= PackageManager.FLAG_PERMISSION_USER_FIXED
-                    | PackageManager.FLAG_PERMISSION_USER_SET;
+                    | PackageManager.FLAG_PERMISSION_USER_SET
+                    | PackageManager.FLAG_PERMISSION_ONE_TIME;
         }
         // If a component gets a permission for being the default handler A and also default handler
         // B, we grant the weaker grant form. This only applies to default permission grant.
@@ -607,7 +608,8 @@ public class Permissions {
         }
         if (!overrideUserSetAndFixed) {
             fixedFlags |= PackageManager.FLAG_PERMISSION_USER_FIXED
-                    | PackageManager.FLAG_PERMISSION_USER_SET;
+                    | PackageManager.FLAG_PERMISSION_USER_SET
+                    | PackageManager.FLAG_PERMISSION_ONE_TIME;
         }
         return (flags & fixedFlags) != 0;
     }
-- 
2.48.1.262.g85cc9f2d1e-goog

